api/envoy/api/v2/srds.proto

```diff
@@ -0,0 +1,153 @@
+  }
+
+  rpc IncrementalScopedRoutes(stream IncrementalDiscoveryRequest)
+      returns (stream IncrementalDiscoveryResponse) {
+  }
+
+  rpc FetchScopedRoutes(DiscoveryRequest) returns (DiscoveryResponse) {
+    option (google.api.http) = {
+      post: "/v2/discovery:scoped-routes"
+      body: "*"
+    };
+  }
+}
+
+// This configuration represents a set of "scopes", each containing independent routing
+// configuration (see :ref:`routing architecture overview <arch_overview_http_routing>`). A scope is
+// assigned to each request based on request attributes, such as the value of a header designated
+// via this configuration.
+// [#comment:next free field: 4]
+message ScopedRouteConfigurationsSet {
+  // The name of the set of route configuration scopes. This will be the
+  // :ref:`scoped_routes_config_set_name
+  // <envoy_api_field_config.filter.network.http_connection_manager.v2.ScopedRds.scoped_routes_config_set_name>`
+  // set in :ref:`envoy_api_msg_config.filter.network.http_connection_manager.v2.ScopedRds`.
+  string name = 1 [(validate.rules).string.min_bytes = 1];
+
+  // Specifies the mechanism for constructing keys based on request attributes to match
+  // :ref:`scopes <envoy_api_field_ScopedRouteConfigurationsSet.scopes>` against.
+  //
+  // Upon receiving a request's headers, the Router will build a key using the algorithm specified
+  // by this message. This key will be used to look up a corresponding
+  // :ref:`Scope <envoy_api_msg_ScopedRouteConfigurationsSet.Scope>` in a table containing the set
+  // of :ref:`scopes <envoy_api_field_ScopedRouteConfigurationsSet.scopes>`.
+  message ScopeKeyBuilder {
+    // Specifies the mechanism for constructing fragments which are composed into keys.
+    message FragmentBuilder {
+      // Specifies how the value of a header should be extracted.
+      // The following example maps the structure of a header to the fields in this message.
+      //
+      // .. code::
+      //
+      //    X-Header: a,b;c,d
+      //    |         || |
+      //    |         || \\----> <element_separator>
+      //    |         ||
+      //    |         |\\----> <element.separator>
+      //    |         |
+      //    |         \\----> <element.key>
+      //    |
+      //    \\----> <name>
+      message HeaderValueExtractor {
+        // The name of the header to extract the value from.
+        string name = 1 [(validate.rules).string.min_bytes = 1];
+
+        // The element separator (e.g., ';' separates 'a;b;c;d').
+        string element_separator = 2;
+
+        // Specifies a key value pair to match on.
+        message KvElement {
+          // The separator between key and value (e.g., '=' separates 'k=v;...').
+          string separator = 1;
+
+          // The key to match on.
+          string key = 2;
+        }
+
+        oneof extract_type {
+          // Specifies the index of the element to extract.
+          int32 index = 3;
```

Comment:
unsigned?

Reply:
I didn't find mention of signed vs unsigned in Envoy's style guide so I defaulted to Google's style of using signed integers and relying on assertions to check val >= 0.